sbi/utils/posterior_ensemble.py

```diff
@@ -0,0 +1,446 @@
+        idxs = torch.multinomial(self._weights, num_samples, replacement=True)
+        samples = []
+        for comp_idx, sample_size in torch.vstack(idxs.unique(return_counts=True)).T:
+            sample_shape_c = torch.Size((sample_size,))
+            samples.append(self.posteriors[comp_idx].sample(sample_shape_c, **kwargs))
+        return torch.vstack(samples).reshape(*sample_shape, -1)
+
+    def log_prob(self, theta: Tensor, individually: bool = False, **kwargs) -> Tensor:
+        r"""Returns the average log-probability of the posterior ensemble
+
+        $\sum_{i}^{N} w_{i} p_i(\theta|x)$.
+
+        All kwargs are passed directly through to `posterior.log_prob()`.
+
+        Args:
+            theta: Parameters $\theta$.
+            individually: If true, returns log weights and log_probs individually.
+
+        Raises:
+            AssertionError if posterior estimators are a mixture of different methods.
+
+        Returns:
+            `(len(θ),)`-shaped average log posterior probability $\log p(\theta|x)$ for
+            θ in the support of the prior, -∞ (corresponding to 0 probability) outside.
+        """
+        assert all(
+            isinstance(posterior, type(self.posteriors[0]))
+            for posterior in self.posteriors
+        ), "`log_prob()` only works for ensembles of the same type of posterior."
+
+        log_probs = torch.stack(
+            [posterior.log_prob(theta, **kwargs) for posterior in self.posteriors]
+        )
+        log_weights = torch.log(self._weights).reshape(-1, 1)
+
+        if individually:
+            return log_weights, log_probs
+        else:
+            return torch.logsumexp(log_weights.expand_as(log_probs) + log_probs, dim=0)
+
+    def set_default_x(self, x: Tensor) -> "NeuralPosteriorEnsemble":
+        r"""Set new default x for `.sample(), .log_prob()` to use as conditioning context.
+
+        This is a pure convenience to avoid having to repeatedly specify `x` in calls to
+        `.sample()` and `.log_prob()` - only θ needs to be passed.
+
+        This convenience is particularly useful when the posterior ensemble is focused,
+        i.e. has been trained over multiple rounds to be accurate in the vicinity of a
+        particular `x=x_o` (you can check if your posterior object is focused by
+        printing one exemplary component of the ensemble).
+
+        NOTE: this method is chainable, i.e. will return the NeuralPosteriorEnsemble
+        object so that calls like `posterior_enemble.set_default_x(my_x).sample(mytheta)
+        ` are possible.
+
+        Args:
+            x: The default observation to set for every posterior $p_i(theta|x)$ in the
+            ensemble.
+        Returns:
+            `NeuralPosteriorEnsemble` that will use a default `x` when not explicitly
+            passed.
+        """
+        for posterior in self.posteriors:
+            posterior.set_default_x(x)
+        self.potential_fn = EnsemblePotentialProvider(
```

Comment:
I don't think there is a reason to run this here. It is run both in `self.potential()` and in `self.map()`.